File: rust/src/contracts.rs
//! Common reference-execution checks. These checks do not provide a sandbox.
use crate::types::{Manifest, ToolCladError};
use crate::validator::validate_arg;
use regex::Regex;
use std::collections::HashMap;

pub const MAX_REQUEST_BYTES: usize = 1024 * 1024;
pub const MAX_RESPONSE_BYTES: usize = 4 * 1024 * 1024;

fn invalid(message: impl Into<String>) -> ToolCladError {
    ToolCladError::ValidationError(message.into())
}

pub(crate) fn value_string(value: &toml::Value) -> String {
    match value {
        toml::Value::String(s) => s.clone(),
        other => other.to_string(),
    }
}

/// Validate supplied arguments and defaults, refusing unknown and reserved names.
pub fn validate_arguments(
    manifest: &Manifest,
    args: &HashMap<String, String>,
) -> Result<HashMap<String, String>, ToolCladError> {
    for name in args.keys() {
        if !manifest.args.contains_key(name) {
            return Err(invalid(format!("unknown argument: {name}")));
        }
    }
    let name_pattern = Regex::new(r"^[A-Za-z][A-Za-z0-9_]*$").unwrap();
    let mut result = HashMap::new();
    for (name, definition) in &manifest.args {
        if !name_pattern.is_match(name) {
            return Err(invalid(format!("invalid argument name: {name}")));
        }
        let fallback_value = definition
            .default
            .as_ref()
            .or_else(|| manifest.command.defaults.as_ref().and_then(|d| d.get(name)));
        if !args.contains_key(name)
            && definition.type_name == "literal_text"
            && fallback_value.is_some_and(|v| !v.is_str())
        {
            return Err(invalid(format!(
                "literal_text default must be a string: {name}"
            )));
        }
        let fallback = fallback_value.map(value_string);
        let value = args.get(name).or(fallback.as_ref());
        let Some(value) = value else {
            if definition.required {
                return Err(invalid(format!("missing required argument: {name}")));
            }
            continue;
        };
        if value.contains('\0')
            || (definition.required
                && definition.type_name != "literal_text"
                && value.trim().is_empty())
        {
            return Err(invalid(format!("invalid empty or NUL argument: {name}")));
        }
        result.insert(name.clone(), validate_arg(name, definition, value)?);
    }
    if result.values().map(String::len).sum::<usize>() > MAX_REQUEST_BYTES {
        return Err(invalid("arguments exceed 1 MiB"));
    }
    Ok(result)
}

pub(crate) fn check_execution(m: &Manifest, dry_run: bool) -> Result<(), ToolCladError> {
    if !(1..=3600).contains(&m.tool.timeout_seconds) {
        return Err(invalid("timeout_seconds must be between 1 and 3600"));
    }
    let backends = [
        m.command.exec.is_some() || m.command.template.is_some(),
        m.command.executor.is_some(),
        m.http.is_some(),
        m.mcp.is_some(),
        m.session.is_some(),
        m.browser.is_some(),
    ];
    if backends.into_iter().filter(|v| *v).count() > 1 {
        return Err(invalid("ambiguous execution backends"));
    }
    if dry_run {
        return Ok(());
    }
    if m.tool.dispatch != "exec"
        || m.tool.human_approval
        || m.tool.cedar.is_some()
        || m.args.values().any(|a| a.scope_check)
    {
        return Err(invalid("execution requires an embedding runtime for dispatch, approval, Cedar or scope enforcement"));
    }
    if let Some(output) = &m.output {
        if !matches!(
            output.parser.as_deref().unwrap_or(""),
            "" | "builtin:json" | "builtin:xml" | "builtin:csv" | "builtin:jsonl" | "builtin:text"
        ) {
            return Err(invalid(
                "custom output parsers require an embedding runtime",
            ));
        }
    }
    Ok(())
}

pub(crate) fn child_environment(cmd: &mut std::process::Command) {
    cmd.env_clear();
    for name in ["PATH", "LANG", "LC_ALL", "SYSTEMROOT"] {
        if let Some(value) = std::env::var_os(name) {
            cmd.env(name, value);
        }
    }
}

pub(crate) fn interpolate(template: &str, args: &HashMap<String, String>) -> String {
    Regex::new(r"\{(\w+)\}")
        .unwrap()
        .replace_all(template, |c: &regex::Captures| {
            args.get(&c[1]).cloned().unwrap_or_else(|| c[0].to_string())
        })
        .into_owned()
}

pub(crate) fn http_url(
    template: &str,
    args: &HashMap<String, String>,
) -> Result<String, ToolCladError> {
    let authority = Regex::new(r"^(https?)://([^/?#]+)").unwrap();
    let matched = authority
        .captures(template)
        .ok_or_else(|| invalid("HTTP URL requires a fixed http(s) authority"))?;
    if matched[2].contains(['{', '}', '@', '\\']) || template.contains("{_secret:") {
        return Err(invalid(
            "HTTP URL requires a fixed authority without credentials or secrets",
        ));
    }
    let value = interpolate(template, args);
    let parsed = reqwest::Url::parse(&value).map_err(|_| invalid("invalid HTTP URL"))?;
    if authority.find(&value).map(|m| m.as_str()) != Some(&matched[0])
        || parsed.fragment().is_some()
        || value.chars().any(|c| c <= ' ' || c == '\\')
    {
        return Err(invalid("invalid HTTP URL or changed authority"));
    }
    Ok(value)
}

pub(crate) fn http_template(
    template: &str,
    args: &HashMap<String, String>,
    dry_run: bool,
    json_string: bool,
) -> Result<String, ToolCladError> {
    let mut failure = None;
    let value = Regex::new(r"\{(_secret:[A-Za-z0-9_]+|\w+)\}")
        .unwrap()
        .replace_all(template, |c: &regex::Captures| {
            let value = if let Some(name) = c[1].strip_prefix("_secret:") {
                if dry_run {
                    "[REDACTED]".to_string()
                } else {
                    let key = format!("TOOLCLAD_SECRET_{}", name.to_uppercase());
                    std::env::var(&key).unwrap_or_else(|_| {
                        failure = Some(invalid(format!(
                            "missing secret environment variable: {key}"
                        )));
                        String::new()
                    })
                }
            } else {
                args.get(&c[1]).cloned().unwrap_or_else(|| c[0].to_string())
            };
            if json_string {
                let encoded = serde_json::to_string(&value).unwrap();
                encoded[1..encoded.len() - 1].to_string()
            } else {
                value
            }
        })
        .into_owned();
    if let Some(error) = failure {
        return Err(error);
    }
    Ok(value)
}

File: js/src/contracts.js
/** Common reference-execution checks; these do not provide a sandbox. */
import { validateArg } from './validator.js';
import http from 'node:http';
import https from 'node:https';

export const MAX_REQUEST_BYTES = 1024 * 1024;
export const MAX_RESPONSE_BYTES = 4 * 1024 * 1024;
const token = /\{(_secret:[A-Za-z0-9_]+|\w+)\}/g;

export function validateArguments(manifest, args) {
  const definitions = manifest.args || {};
  for (const name of Object.keys(args)) {
    if (!Object.hasOwn(definitions, name)) throw new Error(`Unknown argument: ${name}`);
  }
  const resolved = {};
  for (const [name, def] of Object.entries(definitions)) {
    if (!/^[A-Za-z][A-Za-z0-9_]*$/.test(name) || ['constructor', 'prototype', '__proto__'].includes(name)) {
      throw new Error(`Invalid argument name: ${name}`);
    }
    const value = Object.hasOwn(args, name) ? args[name] : (def.default ?? manifest.command?.defaults?.[name]);
    if (def.type === "literal_text" && Object.hasOwn(args, name) && typeof value !== "string") throw new Error(`literal_text requires a string: ${name}`);
    if (value === undefined || value === null) {
      if (def.required) throw new Error(`Missing required argument: ${name}`);
      continue;
    }
    if (String(value).includes('\0') || (def.required && def.type !== "literal_text" && !String(value).trim())) throw new Error(`Invalid empty or NUL argument: ${name}`);
    resolved[name] = validateArg(def, value);
  }
  if (Object.values(resolved).reduce((n, v) => n + Buffer.byteLength(String(v)), 0) > MAX_REQUEST_BYTES) throw new Error('Arguments exceed 1 MiB');
  return resolved;
}

export function checkExecution(manifest, dryRun = false) {
  const timeout = manifest.tool.timeout_seconds ?? 60;
  if (!Number.isInteger(timeout) || timeout < 1 || timeout > 3600) throw new Error('timeout_seconds must be between 1 and 3600');
  const command = manifest.command || {};
  if ([command.exec?.length || command.template, command.executor, manifest.http, manifest.mcp, manifest.session, manifest.browser].filter(Boolean).length > 1) throw new Error('Ambiguous execution backends');
  if (dryRun) return;
  if (Object.hasOwn(manifest, 'filesystem')) throw new Error('Filesystem grants require an embedding runtime');
  if ((manifest.tool.dispatch ?? 'exec') !== 'exec' || manifest.tool.human_approval || manifest.tool.cedar || Object.values(manifest.args || {}).some(a => a.scope_check)) {
    throw new Error('Execution requires an embedding runtime for dispatch, approval, Cedar or scope enforcement');
  }
  const parser = manifest.output?.parser;
  if (parser && !['builtin:json', 'builtin:xml', 'builtin:csv', 'builtin:jsonl', 'builtin:text'].includes(parser)) throw new Error('Custom output parsers require an embedding runtime');
}

export function childEnvironment() {
  return Object.fromEntries(['PATH', 'LANG', 'LC_ALL', 'SYSTEMROOT'].filter(k => process.env[k] !== undefined).map(k => [k, process.env[k]]));
}

export function splitTemplate(text) {
  const parts = [];
  let current = '', quote = '', started = false;
  for (let i = 0; i < text.length; i++) {
    const c = text[i];
    if (c === '\\' && quote !== "'") {
      const next = text[++i];
      if (next === undefined) throw new Error('Unfinished command escape');
      if (quote === '"' && !['"', '\\', '$', '`', '\n'].includes(next)) current += '\\';
      if (next !== '\n') current += next;
      started = true;
    } else if (quote) {
      if (c === quote) quote = ''; else current += c;
    } else if (c === "'" || c === '"') {
      quote = c; started = true;
    } else if (/\s/.test(c)) {
      if (started) { parts.push(current); current = ''; started = false; }
    } else { current += c; started = true; }
  }
  if (quote) throw new Error('Unclosed command quote');
  if (started) parts.push(current);
  return parts;
}

export function quoteArg(value) {
  const text = String(value);
  return /^[A-Za-z0-9_@%+=:,./-]+$/.test(text) ? text : "'" + text.replaceAll("'", "'\"'\"'") + "'";
}

export function templateArgv(template, values, fragments, present = {}) {
  const substitute = t => t.replace(token, (m, k) => String(values[k] ?? m));
  const argv = [];
  for (const part of splitTemplate(template)) {
    const match = /^\{(\w+)\}$/.exec(part);
    if (match && Object.hasOwn(fragments, match[1])) argv.push(...splitTemplate(fragments[match[1]]).map(substitute));
    else { const value = substitute(part); if (value || !part || [...part.matchAll(token)].some(m => Object.hasOwn(present, m[1]))) argv.push(value); }
  }
  if (!argv.length || !argv[0]) throw new Error('Command produced empty argv');
  return argv;
}

export function prepareHttp(manifest, args, dryRun = false) {
  args = validateArguments(manifest, args);
  checkExecution(manifest, dryRun);
  const def = manifest.http;
  const authority = /^(https?):\/\/([^/?#]+)/.exec(def.url);
  if (!authority || /[{}@\\]/.test(authority[2]) || def.url.includes('{_secret:')) throw new Error('HTTP URL requires a fixed http(s) authority without credentials or secrets');
  const url = def.url.replace(token, (m, k) => String(args[k] ?? m));
  const parsed = new URL(url);
  if (/[\x00-\x20\\]/.test(url) || parsed.hash || /^(https?):\/\/([^/?#]+)/.exec(url)?.[0] !== authority[0]) throw new Error('Invalid HTTP URL or changed authority');
  function render(template, json = false) {
    return template.replace(token, (m, key) => {
      let value;
      if (key.startsWith('_secret:')) {
        const name = 'TOOLCLAD_SECRET_' + key.slice(8).toUpperCase();
        value = dryRun ? '[REDACTED]' : process.env[name];
        if (value === undefined) throw new Error(`Missing secret environment variable: ${name}`);
      } else value = args[key] ?? m;
      return json ? JSON.stringify(String(value)).slice(1, -1) : String(value);
    });
  }
  const headers = Object.fromEntries(Object.entries(def.headers || {}).map(([k, v]) => [k, render(v)]));
  for (const [k, v] of Object.entries(headers)) if (/[\r\n\0]/.test(k + v)) throw new Error('Invalid HTTP header');
  const body = def.body_template === undefined ? undefined : render(def.body_template, true);
  if (Buffer.byteLength(url + (body || '') + Object.entries(headers).flat().join('')) > MAX_REQUEST_BYTES) throw new Error('HTTP request exceeds 1 MiB');
  return { url, headers, body, resolvedArgs: args };
}

export function requestHttp(url, options, timeoutSeconds) {
  // Core HTTP clients neither redirect nor consult proxy environment variables.
  return new Promise((resolve, reject) => {
    const client = new URL(url).protocol === 'https:' ? https : http;
    let timer;
    const req = client.request(url, { ...options, agent: false }, res => {
      const chunks = []; let size = 0;
      res.on('data', chunk => {
        size += chunk.length;
        if (size > MAX_RESPONSE_BYTES) req.destroy(new Error('HTTP response exceeds 4 MiB'));
        else chunks.push(chunk);
      });
      res.on('error', reject);
      res.on('end', () => { clearTimeout(timer); resolve({ status: res.statusCode, body: Buffer.concat(chunks).toString('utf8') }); });
    });
    req.on('error', e => { clearTimeout(timer); reject(e); });
    timer = setTimeout(() => req.destroy(new Error('HTTP request timed out')), timeoutSeconds * 1000);
    req.end(options.body);
  });
}
